### PR TITLE
fixes #10 #11 #12

### LINKED ISSUE #10: unitless quantities cannot be printed

### PR DESCRIPTION
**Pull request by apmccartney**

### CODE REVIEW

File: src/dimwits/constants.hpp [hunk 1]
```diff
@@ -1,5 +1,6 @@
 namespace constant {
 
+constexpr auto unitless = 1.0 * meter / meter;
```
Comment (jlconlin) [inline src/dimwits/constants.hpp:3]: I think this should be `gram/gram`

File: src/dimwits/quantity/operation/src/assignment.hpp [hunk 2]
```diff
@@ -31,3 +31,19 @@ quantity::Type< UnitLeft, MagnitudeLeft >& operator-=
   left.value -= conversion::factor<UnitRight, UnitLeft> * right.value;
   return left;
 }
+
+template< typename Magnitude, typename Unit, typename Factor >
+auto operator*=
+( quantity::Type< Unit, Magnitude >& left, const Factor& right )
+  -> std::conditional_t< true, decltype(left), decltype( left = left * right ) >{
```
Comment (jlconlin) [inline src/dimwits/quantity/operation/src/assignment.hpp:38]: I understand what this is doing (ensuring that the types are appropriate), but I don't understand that `->` syntax. Let's talk about it.

File: src/dimwits/quantity/test/operations.test.cpp [hunk 3]
```diff
@@ -114,6 +114,44 @@ TEST_CASE("operations"){
...
+
+      auto generate_unit = decltype(length)::Units();
+      REQUIRE( generate_unit == Centi<Meter>{} );
+    }
+    SECTION("unitless"){
+      quantity::Type< Centi<Meter> > length = 1.0 * meter;
+      length *= 5.0 * meter / meter;
+      REQUIRE( 500.0 == length.value );
+
+      auto generate_unit = decltype(length)::Units();
```
Comment (jlconlin) [inline src/dimwits/quantity/test/operations.test.cpp:132]: Is this how you extract the unit from a `Quantity`?

File: src/dimwits/quantity/test/operations.test.cpp [hunk 4]
```diff
@@ -114,6 +114,44 @@ TEST_CASE("operations"){
...
+      length *= 5.0;
+      REQUIRE( 500.0 == length.value );
+
+      auto generate_unit = decltype(length)::Units();
+      REQUIRE( generate_unit == Centi<Meter>{} );
+    }
+    SECTION("unitless"){
+      quantity::Type< Centi<Meter> > length = 1.0 * meter;
+      length *= 5.0 * meter / meter;
+      REQUIRE( 500.0 == length.value );
```
Comment (jlconlin) [inline src/dimwits/quantity/test/operations.test.cpp:130]: This looks like it should fail, until you realize that it is centimeters and not meters.

File: src/dimwits/quantity/test/operations.test.cpp [hunk 5]
```diff
@@ -173,13 +211,21 @@ TEST_CASE("operations"){
...
-    ss << length0 << length1 << length2 << (length0 * length1 * length2);
-    REQUIRE( ss.str() == " 1 ft 11 in 1 angstrom 11 angstrom in ft" );
+    {
+      auto length0 = 1.0 * foot;
+      auto length1 = 11.0 * inches;
+      auto length2 = 1.0 * angstroms;
+
+      std::stringstream ss;
+      ss << length0 << length1 << length2 << (length0 * length1 * length2);
+      REQUIRE( ss.str() == " 1 ft 11 in 1 angstrom 11 angstrom in ft" );
```
Comment (jlconlin) [inline src/dimwits/quantity/test/operations.test.cpp:221]: That's a tongue twister!

**Review by jlconlin — APPROVED** [review]
As long as the tests pass, this looks good to me.

File: src/dimwits/quantity/test/operations.test.cpp (hunk 5 above)
Comment (apmccartney) [reply src/dimwits/quantity/test/operations.test.cpp:221]: indeed!

File: src/dimwits/quantity/test/operations.test.cpp (hunk 3 above)
Comment (apmccartney) [reply src/dimwits/quantity/test/operations.test.cpp:132]: How I do it

File: src/dimwits/quantity/test/operations.test.cpp (hunk 4 above)
Comment (apmccartney) [reply src/dimwits/quantity/test/operations.test.cpp:130]: It's good to keep you on  your toes

File: src/dimwits/quantity/operation/src/assignment.hpp (hunk 2 above)
Comment (apmccartney) [reply src/dimwits/quantity/operation/src/assignment.hpp:38]: it's a [trailing return type](https://cppstyle.wordpress.com/trailing-return-type/).

**Pull request merged**